Add CLI test for content-host errata and sub commands - issue #1708

abalakh opened the pull request:
Added 2 CLI tests for ```content-host errata info``` and ```content-host errata apply``` commands.
Resolves #1708
Added setUp section as there're 3 task with same preconditions (#1708, #1709, #1710).
Tasks #1709 and #1710 will need vm with katello-agent but without custom package installed, that's why haven't moved custom package installation to setUp section.
Tests results:
```
nosetests tests/foreman/cli/test_contenthost.py -m test_contenthost_
..
----------------------------------------------------------------------
Ran 2 tests in 844.521s

OK
```

Ichimonji10 commented on robottelo/vm.py:
```diff
@@ -16,7 +16,7 @@
 
 
-BASE_IMAGES = ('rhel7', 'rhel65', 'rhel66')
+BASE_IMAGES = ('rhel71', 'rhel70', 'rhel66', 'rhel65')
```
```
[ichimonji10@beech:robottelo]$ git grep rhel7
robottelo/test.py:    vm_os = 'rhel7'
robottelo/vm.py:make sure that the server have in place: the base images for rhel65 and rhel7,
robottelo/vm.py:BASE_IMAGES = ('rhel7', 'rhel65', 'rhel66')
```

abalakh replied:
#1 - good catch. #2 is comment, #3 is actually changed in this PR.
Regarding #1, what's prefered rhel version? The latest one ```rhel71```, or maybe it's better to leave it as it was (```rhel70```) e.g. for compatibility issues?

Ichimonji10 replied:
I think Sat only currently support RHEL 7.1 and RHEL 6.6. Maybe @elyezer knows better?

elyezer replied:
Sat aways support latest RHEL version, in the 7 series is 7.1.

sthirugn replied:
Wait, are you asking about the support for installer or support for content hosts?
for installer - currently for sat 6.1, we support only rhel66, rhel71
for content hosts - we support all, that is - rhel 5.7+, rhel6+, rhel7+
please ping me in irc if you have more questions on this

abalakh replied:
Thanks for clarification, it's clear now.

Ichimonji10 commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -532,3 +553,280 @@ def test_bugzilla_1154611(self):
+        # Enable Red Hat Satellite Tools repo
+        result = self.vm.run(
+            'subscription-manager repos --enable '
+            'rhel-7-server-satellite-tools-6-beta-rpms'
```
Is this repo name likely to change on a frequent basis?

abalakh replied:
I guess it will change just once, after release, from beta-channel to release one. No more changes expected - this repo is only needed to install katello-agent from it. Maybe we should put there some comment like ```# TODO: Change to rhel-7-server-satellite-tools-6-rpms after release```?

Ichimonji10 replied:
Sure. That seems reasonable. The test will break when the repos change, and when someone then goes and looks at the test to fix it, having a TODO will be very handy.

sthirugn replied:
In addition to TODO, perhaps move this repo name to robottelo/robottelo/common/constants.py.  I dont think this will be the only place which needs to install katello-agent.  There may be several coming up.  So moving to constants.py will help to make the future fix in one place.

elyezer replied:
:+1: for @sthirugn suggestion.

Ichimonji10 commented:
ACK pending comments.

elyezer commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -532,3 +553,280 @@ def test_bugzilla_1154611(self):
+        self.assertIn('katello-agent', result.stdout[0])
+
+    def tearDown(self):
+        self.vm.destroy()
```
:+1:

elyezer commented:
ACK, the tests looks good to me. Maybe @sthirugn will want to check the steps on the test, as he have worked more with errata then maybe have any suggestion.

sthirugn commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -532,3 +553,280 @@ def test_bugzilla_1154611(self):
+        """
+        super(TestContentHostKatelloAgent, self).setUp()
+        # Product, RepoSet and repository variables
+        rhel_product_name = 'Red Hat Enterprise Linux Server'
```
Move this to constants.py

sthirugn commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -532,3 +553,280 @@ def test_bugzilla_1154611(self):
+        rhel_product_name = 'Red Hat Enterprise Linux Server'
+        rhel_repo_set = (
+            'Red Hat Satellite Tools 6 Beta '
+            '(for RHEL 7 Server) (RPMs)'
```
Move this to constants.py

sthirugn commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -532,3 +553,280 @@ def test_bugzilla_1154611(self):
+        rhel_repo_name = (
+            'Red Hat Satellite Tools 6 Beta '
+            'for RHEL 7 Server '
+            'RPMs x86_64 7Server'
```
Move this to constants.py

sthirugn commented on tests/foreman/cli/test_contenthost.py:
```diff
@@ -532,3 +553,280 @@ def test_bugzilla_1154611(self):
+        result = self.vm.run('rpm -i {0}.rpm'.format(FAKE_0_CUSTOM_PACKAGE))
+        self.assertEqual(result.return_code, 0)
+        # Apply errata to content host
+        result = ContentHost.errata_apply({
```
I believe you are applying this in synchronous way.  Can you assert in the vm to make sure it got the errata?

abalakh replied:
No, it's async operation - it just schedules errata for installation. Only message and ```return_code``` are returned, so nothing more to verify here.

sthirugn replied:
ok

sthirugn commented:
This is a great test.  But I like this to be modularised so the code can be reused in other future tests.

I feel most of the katello agent related code should be contained in the robottelo.cli.contenthost module itself for easy reuse.  From the initial review, I feel that the following code that can be modularized to individual functions:

* Download and Install katello-ca rpm -> robottelo.cli.contenthost.install_katello_cert()
* Register client with foreman server using activation-key -> robottelo.cli.contenthost.register()
* Enable Red Hat Satellite Tools repo or any given repo -> robottelo.cli.contenthost.enable_repo()
* Install katello-agent package -> robottelo.cli.install_katello_agent()

Also the following code in setup() can be modularized.  If you can generalize the code corresponding to the following and build a function called setup_org_for_content_host_tests() may be in robottelo.cli.contenthost, it will be great and can be reused from any other tests. 
* Create new org and environment
* Clone manifest and upload it
* Enable repo from Repository Set
* Synchronize the RH repository
* Create CV and associate repo with it
* Publish a version1 of CV
* Promote version1 to next env
* Create activation key
* Add subscription to activation-key

I am trying to get us to a state where I can easily create content related tests like:
Note: Please excuse syntax or run time errors, I am just trying to give an example of what I am trying to explain in the above paragraphs.
```py
from robottelo.cli.contenthost import ContentHost
from robottelo.vm import VirtualMachine
ContentHost.setup_org_for_content_host_tests() # we can pass the required parameters (may be org?) and get back the required parameters as needed
vm = VirtualMachine(distro='rhel71')
vm.create()
ContentHost.install_katello_cert(vm)
register(vm, activationkey)
enable_repo(vm, repo)
install_katello_agent(vm)
....
....
ContentHost.install_package(vm) #future
ContentHost.remove_package(vm) #future
ContentHost.install_package_group(vm) #future
ContentHost.remove_package_group(vm) #future
ContentHost.install_errata(vm) #future
ContentHost.remove_errata(vm) #future
```

abalakh commented:
Rebased and updated according to comments.
Test results after update:
```
nosetests tests/foreman/cli/test_contenthost.py -m test_contenthost_
..
----------------------------------------------------------------------
Ran 2 tests in 737.574s

OK
```
Regarding @sthirugn suggestion, I think that's a great idea, and it will help in writing next tests, but this suggestion is out of scope of current task and will require refactoring of literally all the code in this PR. Maybe I'll better create separate github issue for modularizing the code and start working on it just after this PR is closed?

sthirugn commented:
@abalakh I dont understand why it is out of scope for this PR

abalakh commented:
@sthirugn sure it is. I meant, the original task (issue #1708) was to add coverage for 2 CLI commands, not to create all the required functions for errata testing, so I suggested to split this task to 2 parts. But if it's neccessary i'll do it in scope of this task, whatever approach is better for you :)

sthirugn commented:
@abalakh Thank you, I created https://github.com/SatelliteQE/robottelo/issues/2199 for you.

sthirugn commented:
ACK

Ichimonji10 commented on robottelo/test.py:
```diff
@@ -291,7 +291,7 @@ def tearDownClass(cls):  # noqa
     vm_cpu = 2
     vm_ram = 6144
-    vm_os = 'rhel7'
+    vm_os = 'rhel71'
```
:+1:

Ichimonji10 commented:
Thanks for reviewing this and creating #2199, @sthirugn. This PR looks good to me.
